src/DescriptorsCommand.cc

```diff
@@ -409,94 +412,240 @@ void AddDescriptor::retrieve_aws_descriptorSet(const std::string &set_path) {
-      std::cout << "Deleted " << n << " files or directories\n";
+        retrieve_aws_descriptorSet(set_path);
     }
 
-    return -1;
-  }
+    // Note dimensionse are based on a 32 bit integer, hence the /4 math on size
+    // as the string blob is sized in 8 bit ints.
+    nr_expected_descs = prop_list.size();
+    expected_blb_size = nr_expected_descs * dimensions * 4;
+
+    //Verify length of input is matching expectations
+    if (blob.length() != expected_blb_size) {
+        std::cerr << "AddDescriptor::insert_descriptor: ";
+        std::cerr << "Expectected Blob Length Does Not Match Input ";
+        std::cerr << blob.length() << " != " << expected_blb_size << std::endl;
+        error["info"] = "FV Input Length Mismatch";
+        return -1;
+    }
 
-  props[VDMS_DESC_ID_PROP] = Json::Int64(id);
+    //TODO modify insert descriptor to handle batches
+    long id = insert_descriptor(blob, set_path, nr_expected_descs, label, error);
 
-  int node_ref = get_value<int>(cmd, "_ref", query.get_available_reference());
+    if (id < 0) {
+        error["status"] = RSCommand::Error;
 
-  query.AddNode(node_ref, VDMS_DESC_TAG, props, Json::nullValue);
+        if (_use_aws_storage) {
+            // delete files in set_path
+            std::uintmax_t n = fs::remove_all(set_path);
+            std::cout << "Deleted " << n << " files or directories\n";
+        }
 
-  // It passed the checker, so it exists.
-  int set_ref = query.get_available_reference();
+        return -1;
+    }
 
-  Json::Value link;
-  Json::Value results;
-  Json::Value list_arr;
-  list_arr.append(VDMS_DESC_SET_PATH_PROP);
-  list_arr.append(VDMS_DESC_SET_DIM_PROP);
-  results["list"] = list_arr;
+    //get reference tag for source node for ID
+    // Loop over properties list, add relevant query, link, and edges for each
+    for(int i=0; i < nr_expected_descs; i++) {
+        int node_ref = query.get_available_reference();
+        Json::Value cur_props;
+        cur_props = prop_list[i];
+        //TODO Note using iterator to modify ID return, we're gonna want to watch this closely.
+        cur_props[VDMS_DESC_ID_PROP] = Json::Int64(id+i);
+        cur_props[VDMS_DESC_LABEL_PROP] = label;
+
+        query.AddNode(node_ref, VDMS_DESC_TAG, cur_props, Json::nullValue);
+
+        // It passed the checker, so it exists.
+        int set_ref = query.get_available_reference();
+
+        Json::Value link;
+        Json::Value results;
+        Json::Value list_arr;
+        list_arr.append(VDMS_DESC_SET_PATH_PROP);
+        list_arr.append(VDMS_DESC_SET_DIM_PROP);
+        results["list"] = list_arr;
+
+        //constraints for getting set node to link to.
+        Json::Value constraints;
+        Json::Value name_arr;
+        name_arr.append("==");
+        name_arr.append(set_name);
+        constraints[VDMS_DESC_SET_NAME_PROP] = name_arr;
+
+        bool unique = true;
+
+        // Query set node-We only need to do this once, outside of the loop TODO MOVE
```

Comment:
The comment in the code makes sense. Why get the reference to the same set per iteration, when the set we want to refer to is the same.
Moving the related code (L590-L613) out of the loop would simplify the operations in the loop to just an AddNode, add_link, AddEdge. (Probably will reduce the overall insert time too as we avoid the queries to get DescriptorSet reference for every element in the batch!)

Reply:
Yup, was a mistake there, even had the TODO (oops!) , testing this now before pushing change.